tasks/ta_finisher.py

```diff
@@ -0,0 +1,649 @@
+
+        lock_manager = LockManager(
+            repoid=repoid,
+            commitid=commitid,
+            report_type=ReportType.COVERAGE,
+            lock_timeout=max(80, self.hard_time_limit_task),
+        )
+
+        try:
+            # this needs to be the coverage notification lock
+            # since both tests post/edit the same comment
+            with lock_manager.locked(
+                LockType.NOTIFICATION,
+                retry_num=self.request.retries,
+            ):
+                finisher_result = self.process_impl_within_lock(
+                    db_session=db_session,
+                    repoid=repoid,
+                    commitid=commitid,
+                    commit_yaml=UserYaml.from_dict(commit_yaml),
+                    previous_result=chord_result,
+                    **kwargs,
+                )
+            if finisher_result["queue_notify"]:
+                self.app.tasks[notify_task_name].apply_async(
+                    args=None,
+                    kwargs=dict(
+                        repoid=repoid,
+                        commitid=commitid,
+                        current_yaml=commit_yaml,
+                    ),
+                )
+
+            return finisher_result
+
+        except LockRetry as retry:
+            self.retry(max_retries=5, countdown=retry.countdown)
+
+    def process_impl_within_lock(
+        self,
+        *,
+        db_session: Session,
+        repoid: int,
+        commitid: str,
+        commit_yaml: UserYaml,
+        previous_result: list[bool],
+        **kwargs,
+    ):
+        log.info("Running test results finishers", extra=self.extra_dict)
+        TestResultsFlow.log(TestResultsFlow.TEST_RESULTS_FINISHER_BEGIN)
+
+        commit: Commit = (
+            db_session.query(Commit).filter_by(repoid=repoid, commitid=commitid).first()
+        )
+        assert commit, "commit not found"
+        repo = commit.repository
+
+        uploads = (
+            db_session.query(Upload)
+            .join(CommitReport)
+            .filter(
+                CommitReport.commit_id == commit.id,
+                CommitReport.report_type == ReportType.TEST_RESULTS.value,
+            )
+            .all()
+        )
+
+        redis_client = get_redis_connection()
+
+        tests_to_write: dict[str, dict[str, Any]] = {}
+        test_instances_to_write: list[dict[str, Any]] = []
+        daily_totals: dict[str, DailyTotals] = dict()
+        test_flag_bridge_data: list[dict] = []
+
+        repo_flakes: list[Flake] = (
+            db_session.query(Flake.testid)
+            .filter(Flake.repoid == repoid, Flake.end_date.is_(None))
+            .all()
+        )
+
+        flaky_test_set = {flake.testid for flake in repo_flakes}
+
+        for upload in uploads:
+            repo_flag_ids = get_repo_flag_ids(db_session, repoid, upload.flag_names)
+            if upload.state == "processed":
```

Comment:
this loop has a couple of problems:
- you are querying *all* uploads from the DB, but only ever run the code on `processed` ones
- you only append to `tests_to_write` and friends, but never clear those across uploads
- `save_tests` and friends runs for *all* the uploads, together with never clearing `tests_to_write` above means that you insert the same tests over and over again depending on how many total uploads you have
- you unconditionally set `state = "finished"` for *all* the downloads, also ones that already have that state
- the intermediate msgpack file is never cleared.